### PR TITLE
Fixes hot restart correctly reset the browser history

### PR DESCRIPTION
**Pull request by chunhtai**
web engine hot restart does not tear down the browser history. When the app restarted, it confused itself with the existing browser history, thus causes error in https://github.com/flutter/flutter/issues/73793.

fixes https://github.com/flutter/flutter/issues/73793

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.
- [ ] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

File: lib/web_ui/lib/src/engine.dart [hunk 1]
```diff
@@ -203,6 +203,7 @@ void initializeEngine() {
   // This extension does not need to clean-up Dart statics. Those are cleaned
   // up by the compiler.
   developer.registerExtension('ext.flutter.disassemble', (_, __) {
+    window.resetHistory();
```
Comment (chunhtai) [inline lib/web_ui/lib/src/engine.dart:206]: Hi @mdebbar is there a way to test this? I couldn't find an example to trigger hard reload during the test.

Comment (mdebbar) [reply lib/web_ui/lib/src/engine.dart:206]: I don't think there's a way to test hot restart yet. Maybe @yjbanov has some thoughts?

Comment (mdebbar) [inline lib/web_ui/lib/src/engine.dart:206]: The pattern that we use in the web engine is to call `registerHotRestartListener` and pass a callback. The callback will be invoked when hot restart is requested.

So somewhere near the code that initializes the history, you could do:
```dart
registerHotRestartListener(() {
  window.resetHistory();
});
```

Example: https://github.com/flutter/engine/blob/42019c8361572ec684e4d2a4d7cc5141394f5855/lib/web_ui/lib/src/engine/platform_dispatcher.dart#L768-L770

**Review by mdebbar — APPROVED** [review]
Thanks for fixing this!

**Pull request merged**